File: ts-implementation/unode/core/ast.ts
import type { Immutable } from './immutable';

export const UNODE_AST_VERSION = '2.0.0-alpha.1' as const;

/** JSON-safe primitive value used throughout the canonical AST. */
export type Primitive = string | number | boolean | null;
export type JsonValue = Primitive | JsonObject | JsonArray;
export type JsonObject = { readonly [key: string]: JsonValue };
export type JsonArray = readonly JsonValue[];
export type StateValue = JsonValue;

/** Expression form used for route params, local bindings, and literals. */
export type UiExpr<T extends Primitive = Primitive> =
	| { readonly kind: 'literal'; readonly value: T }
	| { readonly kind: 'binding'; readonly path: string }
	| { readonly kind: 'param'; readonly name: string };

export type StringOrExpr = string | UiExpr<string>;
export type BoolOrExpr = boolean | UiExpr<boolean>;
export type NumberOrExpr = number | UiExpr<number>;
export type PrimitiveOrExpr = Primitive | UiExpr;

export type Tone = 'default' | 'muted' | 'info' | 'success' | 'warning' | 'danger';
export type Gap = 'none' | 'xs' | 'sm' | 'md' | 'lg';
export type TextRole =
	| 'heading'
	| 'title'
	| 'subtitle'
	| 'body'
	| 'label'
	| 'caption'
	| 'code'
	| 'hint';
export type ContainerRole =
	| 'page'
	| 'panel'
	| 'section'
	| 'group'
	| 'toolbar'
	| 'sidebar'
	| 'dialog';
export type ValueFormat =
	| 'number'
	| 'currency'
	| 'date'
	| 'datetime'
	| 'duration'
	| 'bytes'
	| 'percent'
	| 'raw';
export type ActionIntent = 'primary' | 'secondary' | 'ghost' | 'danger';
export type ActionVariant = 'button' | 'link' | 'icon-button' | 'menu-item';
export type InputKind =
	| 'text'
	| 'textarea'
	| 'number'
	| 'password'
	| 'email'
	| 'url'
	| 'boolean'
	| 'select'
	| 'multiselect'
	| 'date'
	| 'file';
export type AspectRatio = 'square' | 'poster' | 'video' | 'auto';
export type Align = 'start' | 'center' | 'end';
export type GridColumnCount = 1 | 2 | 3 | 4 | 5 | 6;
export type ResponsiveGridColumns = Readonly<{
	base?: GridColumnCount;
	sm?: GridColumnCount;
	md?: GridColumnCount;
	lg?: GridColumnCount;
	xl?: GridColumnCount;
}>;

/** Symbolic action reference interpreted by the runtime or renderer. */
export interface ActionRef {
	readonly type: string;
	readonly params?: Readonly<Record<string, JsonValue>>;
	readonly confirm?: Readonly<{
		title?: StringOrExpr;
		message: StringOrExpr;
	}>;
}

/** Reference to a media resource without committing to a renderer-specific implementation. */
export type MediaRef =
	| { readonly type: 'url'; readonly src: string }
	| { readonly type: 'at-blob'; readonly did: string; readonly cid: string }
	| { readonly type: 'asset'; readonly name: string }
	| {
			readonly type: 'placeholder';
			readonly kind?: 'cover' | 'image' | 'avatar' | 'thumbnail';
			readonly label?: string;
	  };

/**
 * Shared node metadata.
 *
 * `id` serves dual purpose:
 *   - Reconciliation identity: used by renderers to track nodes across
 *     reactive updates, equivalent to React's `key`. Must be unique among
 *     siblings in dynamic collections (lists, grids with continuation).
 *   - Semantic/accessibility identity: becomes the element's DOM id,
 *     aria-labelledby target, and focus restoration anchor.
 *
 * If absent, the normalizer derives a structural fallback from the node's
 * position in the tree (e.g. "screen.c0.c1"). This is stable within a load
 * cycle and safe for static structures. Dynamic collections should always
 * provide explicit ids.
 */
export interface NodeBase {
	readonly id?: string;
	readonly meta?: Readonly<Record<string, JsonValue>>;
}

/** Root screen node returned from plugin `render()`. */
export interface ScreenNode extends NodeBase {
	readonly kind: 'screen';
	readonly title?: StringOrExpr;
	readonly subtitle?: StringOrExpr;
	readonly initialFocus?: string;
	readonly initialState?: Readonly<Record<string, StateValue>>;
	readonly children: readonly UiNode[];
}

/** Semantic section container. */
export interface SectionNode extends NodeBase {
	readonly kind: 'section';
	readonly role?: ContainerRole;
	readonly title?: StringOrExpr;
	readonly description?: StringOrExpr;
	readonly children: readonly UiNode[];
}

/** Vertical layout container. */
export interface StackNode extends NodeBase {
	readonly kind: 'stack';
	readonly gap?: Gap;
	readonly children: readonly UiNode[];
}

/** Horizontal/wrapping layout container. */
export interface InlineNode extends NodeBase {
	readonly kind: 'inline';
	readonly gap?: Gap;
	readonly wrap?: boolean;
	readonly align?: Align;
	readonly children: readonly UiNode[];
}

/** Responsive grid container. */
export interface GridNode extends NodeBase {
	readonly kind: 'grid';
	readonly maxColumns?: GridColumnCount;
	readonly columns?: ResponsiveGridColumns;
	readonly gap?: Gap;
	readonly continuation?: CollectionContinuation;
	readonly children: readonly UiNode[];
}

/** Scrollable container. */
export interface ScrollNode extends NodeBase {
	readonly kind: 'scroll';
	readonly children: readonly UiNode[];
}

/** Semantic text leaf node. */
export interface TextNode extends NodeBase {
	readonly kind: 'text';
	readonly content: StringOrExpr;
	readonly role?: TextRole;
	readonly tone?: Tone;
	readonly emphasis?: 'normal' | 'strong';
	readonly truncate?: boolean;
}

/** Formatted scalar value leaf node. */
export interface ValueNode extends NodeBase {
	readonly kind: 'value';
	readonly value: PrimitiveOrExpr;
	readonly format: ValueFormat;
	readonly currencyCode?: string;
	readonly role?: TextRole;
	readonly tone?: Tone;
}

/** Semantic icon leaf node. */
export interface IconNode extends NodeBase {
	readonly kind: 'icon';
	readonly name: string;
	readonly label: string;
	readonly tone?: Tone;
}

/** Compact badge/chip leaf node. */
export interface BadgeNode extends NodeBase {
	readonly kind: 'badge';
	readonly label: StringOrExpr;
	readonly tone?: Tone;
}

/** Visual separator with optional label. */
export interface DividerNode extends NodeBase {
	readonly kind: 'divider';
	readonly label?: StringOrExpr;
}

/** Media leaf node rendered according to platform capabilities. */
export interface MediaNode extends NodeBase {
	readonly kind: 'media';
	readonly ref: MediaRef;
	readonly mediaKind: 'cover' | 'image' | 'avatar' | 'thumbnail';
	readonly alt: string;
	readonly aspectRatio?: AspectRatio;
	readonly expandable?: boolean;
}

/** Single clickable/focusable region wrapping another node. */
export interface PressableNode extends NodeBase {
	readonly kind: 'pressable';
	readonly child: UiNode;
	readonly action: ActionRef;
	readonly label?: StringOrExpr;
}

/**
 * Reusable collection item row structure.
 *
 * `id` is required on ItemNode because items appear in dynamic collections
 * (ListNode, GridNode) where order may change or items may be inserted/removed.
 * A stable id is the only way the renderer can reconcile items correctly
 * without remounting unchanged rows. Use the underlying data record's natural
 * identifier: work.id, chapter.id, etc.
 */
export interface ItemNode {
	readonly kind: 'item';
	readonly id: string;
	readonly meta?: Readonly<Record<string, JsonValue>>;
	readonly leading?: readonly UiNode[];
	readonly primary: readonly UiNode[];
	readonly secondary?: readonly UiNode[];
	readonly trailing?: readonly UiNode[];
	readonly action?: ActionRef;
}

/** Local continuation strategy that reveals more already-loaded items. */
export interface IncrementalContinuation {
	readonly kind: 'incremental';
	readonly binding: string;
	readonly initial: number;
	readonly step: number;
	readonly label?: StringOrExpr;
}

/** Remote continuation strategy that asks the host/plugin for more items. */
export interface RemoteContinuation {
	readonly kind: 'remote';
	readonly hasMore: boolean;
	readonly loadMore: ActionRef;
	readonly label?: StringOrExpr;
	readonly loadingLabel?: StringOrExpr;
}

/** Supported continuation strategies for large collections. */
export type CollectionContinuation = IncrementalContinuation | RemoteContinuation;

/** Semantic list of collection items. */
export interface ListNode extends NodeBase {
	readonly kind: 'list';
	readonly items: readonly ItemNode[];
	readonly density?: 'compact' | 'normal' | 'comfortable';
	readonly continuation?: CollectionContinuation;
}

/** User-invokable action node. */
export interface ActionNode extends NodeBase {
	readonly kind: 'action';
	readonly label: StringOrExpr;
	readonly action: ActionRef;
	readonly intent?: ActionIntent;
	readonly variant?: ActionVariant;
	readonly leadingIcon?: string;
	readonly disabled?: BoolOrExpr;
}

/** Group of related actions. */
export interface ActionsNode extends NodeBase {
	readonly kind: 'actions';
	readonly align?: Align;
	readonly children: readonly ActionNode[];
}

/** Inline collapsible region controlled by a binding path. */
export interface DisclosureNode extends NodeBase {
	readonly kind: 'disclosure';
	readonly binding: string;
	readonly label: StringOrExpr;
	readonly labelExpanded?: StringOrExpr;
	readonly children: readonly UiNode[];
}

/** Option rendered inside a menu. */
export interface MenuItem {
	readonly id?: string;
	readonly label: StringOrExpr;
	readonly action: ActionRef;
	readonly selected?: boolean;
	readonly disabled?: BoolOrExpr;
}

/** Popup or contextual menu trigger and item list. */
export interface MenuNode extends NodeBase {
	readonly kind: 'menu';
	readonly label: StringOrExpr;
	readonly items: readonly MenuItem[];
	readonly intent?: ActionIntent;
	readonly align?: Exclude<Align, 'center'>;
}

/** Select-like option shape used by certain input kinds. */
export interface SelectOption {
	readonly label: string;
	readonly value: Primitive;
	readonly disabled?: boolean;
}

/** Semantic input field description. */
export interface InputNode extends NodeBase {
	readonly kind: 'input';
	readonly name: string;
	readonly inputKind: InputKind;
	readonly label: StringOrExpr;
	readonly value?: PrimitiveOrExpr;
	readonly placeholder?: StringOrExpr;
	readonly helpText?: StringOrExpr;
	readonly required?: boolean;
	readonly disabled?: BoolOrExpr;
	readonly options?: readonly SelectOption[];
	readonly constraints?: Readonly<{
		min?: number;
		max?: number;
		minLength?: number;
		maxLength?: number;
		pattern?: string;
	}>;
}

/** Semantic form container. */
export interface FormNode extends NodeBase {
	readonly kind: 'form';
	readonly name: string;
	readonly children: readonly UiNode[];
	readonly submit?: ActionRef;
}

/** Status or feedback block. */
export interface StatusNode extends NodeBase {
	readonly kind: 'status';
	readonly severity: Exclude<Tone, 'default' | 'muted'>;
	readonly title?: StringOrExpr;
	readonly message: StringOrExpr;
	readonly actions?: readonly ActionNode[];
}

/** Empty state block for absent content. */
export interface EmptyStateNode extends NodeBase {
	readonly kind: 'empty';
	readonly icon?: string;
	readonly title: StringOrExpr;
	readonly message?: StringOrExpr;
	readonly actions?: readonly ActionNode[];
}

/** Loading indicator block. */
export interface LoadingNode extends NodeBase {
	readonly kind: 'loading';
	readonly label?: StringOrExpr;
	readonly progress?: NumberOrExpr;
}

/** Conditional subtree. */
export interface ConditionalNode extends NodeBase {
	readonly kind: 'conditional';
	readonly condition: BoolOrExpr;
	readonly then: UiNode;
	readonly else?: UiNode;
}

/** Inline slot placeholder with optional fallback UI. */
export interface SlotNode extends NodeBase {
	readonly kind: 'slot';
	readonly name: string;
	readonly fallback?: UiNode;
}

/** All canonical child-capable UI nodes except the root `screen`. */
export type UiNode =
	| SectionNode
	| StackNode
	| InlineNode
	| GridNode
	| ScrollNode
	| TextNode
	| ValueNode
	| IconNode
	| BadgeNode
	| DividerNode
	| MediaNode
	| PressableNode
	| ItemNode
	| ListNode
	| ActionNode
	| ActionsNode
	| DisclosureNode
	| MenuNode
	| InputNode
	| FormNode
	| StatusNode
	| EmptyStateNode
	| LoadingNode
	| ConditionalNode
	| SlotNode;

export type RootNode = ScreenNode | UiNode;

/** String union of all non-root node kinds. */
export type UiNodeKind = UiNode['kind'];
export type ImmutableNode = Immutable<UiNode>;
export type ImmutableScreen = Immutable<ScreenNode>;

File: crates/unode/src/core/ast.rs
//! Canonical, serializable Abstract Syntax Tree for uNode UIs.
use crate::core::chrome::ScreenRouteTabsMeta;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use std::collections::BTreeMap;

pub const UNODE_AST_VERSION: &str = "2.0.0-alpha.1";

// Generic Expression Types

/// Expression form used for route params, local bindings, and literals.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
#[serde(rename_all = "camelCase")]
pub enum UiExpr<T> {
    Literal { value: T },
    Binding { path: String },
    Param { name: String },
}

/// Represents a value that can either be a direct value or a `UiExpr`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum OneOrExpr<T> {
    Value(T),
    Expr(UiExpr<T>),
}

// Type aliases for common OneOrExpr types
pub type StringOrExpr = OneOrExpr<String>;
pub type BoolOrExpr = OneOrExpr<bool>;
pub type NumberOrExpr = OneOrExpr<f64>;
pub type Primitive = Option<JsonValue>; // Represents JSON-safe primitives
pub type PrimitiveOrExpr = OneOrExpr<Primitive>;

// Enumerated Dictionaries

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum Tone {
    Default,
    Muted,
    Info,
    Success,
    Warning,
    Danger,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum Gap {
    None,
    Xs,
    Sm,
    Md,
    Lg,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum TextRole {
    Heading,
    Title,
    Subtitle,
    Body,
    Label,
    Caption,
    Code,
    Hint,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ContainerRole {
    Page,
    Panel,
    Section,
    Group,
    Toolbar,
    Sidebar,
    Dialog,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ValueFormat {
    Number,
    Currency,
    Date,
    Datetime,
    Duration,
    Bytes,
    Percent,
    Raw,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ActionIntent {
    Primary,
    Secondary,
    Ghost,
    Danger,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ActionVariant {
    Button,
    Link,
    IconButton,
    MenuItem,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum InputKind {
    Text,
    Textarea,
    Number,
    Password,
    Email,
    Url,
    Boolean,
    Select,
    Multiselect,
    Date,
    File,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum AspectRatio {
    Square,
    Poster,
    Video,
    Auto,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum Align {
    Start,
    Center,
    End,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum TextEmphasis {
    Normal,
    Strong,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum StatusSeverity {
    Info,
    Success,
    Warning,
    Danger,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ListDensity {
    Compact,
    Normal,
    Comfortable,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum MediaKind {
    Image,
    Video,
    Audio,
    Cover,
    Avatar,
    Thumbnail,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum ActionType {
    Core(CoreActionType),
    Custom(String),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum CoreActionType {
    Navigate,
    Submit,
    Dismiss,
    Refresh,
    LoadMore,
}

// Core AST Structures

/// Shared node metadata.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct NodeBase {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<BTreeMap<String, JsonValue>>,
}

/// Symbolic action reference interpreted by the runtime or renderer.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ActionRef {
    pub r#type: ActionType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<BTreeMap<String, JsonValue>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confirm: Option<ActionConfirm>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ActionConfirm {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<StringOrExpr>,
    pub message: StringOrExpr,
}

/// Reference to a media resource.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type")]
#[serde(rename_all = "camelCase")]
pub enum MediaRef {
    Url {
        src: String,
    },
    #[serde(rename = "at-blob")]
    AtBlob {
        did: String,
        cid: String,
    },
    Asset {
        name: String,
    },
    Placeholder {
        #[serde(skip_serializing_if = "Option::is_none")]
        kind: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        label: Option<String>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct ResponsiveGridColumns {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sm: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub md: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lg: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub xl: Option<u8>,
}

// UI Node Definitions

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
#[serde(rename_all = "camelCase")]
pub enum UiNode {
    Section(SectionNode),
    Stack(StackNode),
    Inline(InlineNode),
    Grid(GridNode),
    Scroll(ScrollNode),
    Text(TextNode),
    Value(ValueNode),
    Icon(IconNode),
    Badge(BadgeNode),
    Divider(DividerNode),
    Media(MediaNode),
    Pressable(PressableNode),
    Item(ItemNode),
    List(ListNode),
    Action(ActionNode),
    Actions(ActionsNode),
    Disclosure(DisclosureNode),
    Menu(MenuNode),
    Input(InputNode),
    Form(FormNode),
    Status(StatusNode),
    Empty(EmptyStateNode),
    Loading(LoadingNode),
    Conditional(ConditionalNode),
    Slot(SlotNode),
}

/// Root screen node returned from plugin `render()`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ScreenNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subtitle: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub route_tabs: Option<ScreenRouteTabsMeta>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub initial_focus: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub initial_state: Option<BTreeMap<String, JsonValue>>,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SectionNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<ContainerRole>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<StringOrExpr>,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct StackNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gap: Option<Gap>,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct InlineNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gap: Option<Gap>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wrap: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub align: Option<Align>,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct GridNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_columns: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub columns: Option<ResponsiveGridColumns>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gap: Option<Gap>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub continuation: Option<CollectionContinuation>,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ScrollNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct TextNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub content: StringOrExpr,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<TextRole>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tone: Option<Tone>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emphasis: Option<TextEmphasis>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub truncate: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ValueNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub value: PrimitiveOrExpr,
    pub format: ValueFormat,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub currency_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<TextRole>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tone: Option<Tone>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct IconNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub name: String,
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tone: Option<Tone>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct BadgeNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub label: StringOrExpr,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tone: Option<Tone>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DividerNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<StringOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MediaNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub r#ref: MediaRef,
    pub media_kind: MediaKind,
    pub alt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aspect_ratio: Option<AspectRatio>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expandable: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PressableNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub child: Box<UiNode>,
    pub action: ActionRef,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<StringOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ItemNode {
    // Note: No `base` field. `id` is mandatory and not part of `NodeBase`.
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<BTreeMap<String, JsonValue>>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub leading: Vec<UiNode>,
    pub primary: Vec<UiNode>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub secondary: Vec<UiNode>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub trailing: Vec<UiNode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub action: Option<ActionRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
#[serde(rename_all = "camelCase")]
pub enum CollectionContinuation {
    Incremental(IncrementalContinuation),
    Remote(RemoteContinuation),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct IncrementalContinuation {
    pub binding: String,
    pub initial: u32,
    pub step: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<StringOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RemoteContinuation {
    pub has_more: bool,
    pub load_more: ActionRef,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub loading_label: Option<StringOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ListNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub items: Vec<ItemNode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub density: Option<ListDensity>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub continuation: Option<CollectionContinuation>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ActionNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub label: StringOrExpr,
    pub action: ActionRef,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub intent: Option<ActionIntent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variant: Option<ActionVariant>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub leading_icon: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<BoolOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ActionsNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub align: Option<Align>,
    pub children: Vec<ActionNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DisclosureNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub binding: String,
    pub label: StringOrExpr,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label_expanded: Option<StringOrExpr>,
    pub children: Vec<UiNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MenuItem {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub label: StringOrExpr,
    pub action: ActionRef,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<BoolOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MenuNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub label: StringOrExpr,
    pub items: Vec<MenuItem>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub intent: Option<ActionIntent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub align: Option<Align>, // Exclude "center"
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SelectChoice {
    pub label: String,
    pub value: Primitive,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct InputConstraints {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_length: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_length: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pattern: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct InputNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub name: String,
    pub input_kind: InputKind,
    pub label: StringOrExpr,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<PrimitiveOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub placeholder: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub help_text: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disabled: Option<BoolOrExpr>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub options: Vec<SelectChoice>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub constraints: Option<InputConstraints>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FormNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub name: String,
    pub children: Vec<UiNode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub submit: Option<ActionRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct StatusNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub severity: StatusSeverity,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<StringOrExpr>,
    pub message: StringOrExpr,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<ActionNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EmptyStateNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
    pub title: StringOrExpr,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<StringOrExpr>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub actions: Vec<ActionNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct LoadingNode {
    #[serde(flatten)]
    pub base: NodeBase,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<StringOrExpr>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub progress: Option<NumberOrExpr>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConditionalNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub condition: BoolOrExpr,
    pub r#then: Box<UiNode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#else: Option<Box<UiNode>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SlotNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fallback: Option<Box<UiNode>>,
}
